### ISSUE TITLE
SAML problems: invalid 'xs:ID', and missing 'protocolSupportEnumeration'

### ISSUE THREAD

**Issue by user1970869** [FusionAuth/fusionauth-issues#235]
### Description 

Trouble getting FusionAuth as IDP to pass samltest.id. Identical problems encountered with separate Shibboleth SP.

### Steps to reproduce

1. Setup FusionAuth
2. Add application
3. Setup SAML with info from [1]
4. View application and copy SAML metadata URL to [2]
5. Get 2 errors

moment.metadata:1: element EntityDescriptor: Schemas validity error : Element '{urn:oasis:names:tc:SAML:2.0:metadata}EntityDescriptor', attribute 'ID': '64643134-3530-3365-6433-393236336261' is not a valid value of the atomic type 'xs:ID'.

moment.metadata:1: element IDPSSODescriptor: Schemas validity error : Element '{urn:oasis:names:tc:SAML:2.0:metadata}IDPSSODescriptor': The attribute 'protocolSupportEnumeration' is required but missing. moment.metadata fails to validate

### Platform

 - Ubuntu 18.04
 - NGINX reverse proxy

### Possible source of first error

As per [3], prepending ID in XML metadata with _ (underscore) resolves the first error.

### Additional context

[1] https://samltest.id/download/#SAMLtest_Metadata

[2] https://samltest.id/upload.php

[3] https://docs.secureauth.com/pages/viewpage.action?pageId=6226279

**Comment by robotdan** [comment]
I've got a test fix, I'm going to do some more verification, and then if you're interested I can give you a test.

**Comment by user1970869** [comment]
Indeed interested. Thank you.

Aside, we got everything working by, as mentioned previously, by manually editing the metadata provided to the SP:

1. Prepending the ID with _ (underscore) to resolve the first error, and

2. adding in aline with

`<ns2:IDPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">`

just after the ID section to resolve the second error.

Looking forward to a more permanent fix to the generated metadata.

**Comment by robotdan** [comment]
BTW - thanks for the great reproduce steps, always makes it easier to debug! 

I was able use your steps and recreate the following errors as you've documented as well. 


>1: element EntityDescriptor: Schemas validity error : Element '{urn:oasis:names:tc:SAML:2.0:metadata}EntityDescriptor', attribute 'ID': '39636130-3063-3739-6235-363237353732' is not a valid value of the atomic type 'xs:ID'.
>
> 2: element IDPSSODescriptor: Schemas validity error : Element '{urn:oasis:names:tc:SAML:2.0:metadata}IDPSSODescriptor': The attribute 'protocolSupportEnumeration' is required but missing.


With a patch, I produced the following metadata which passed the validation w/out error. 

```xml
<ns2:EntityDescriptor xmlns:ns2="urn:oasis:names:tc:SAML:2.0:metadata" xmlns="http://www.w3.org/2000/09/xmldsig#" entityID="https://local.fusionauth.io/39636130-3063-3739-6235-363237353732" ID="_39636130-3063-3739-6235-363237353732">
  <ns2:IDPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <ns2:KeyDescriptor use="signing">
      <KeyInfo>
        <X509Data>
          <X509Certificate>
            MIICuDCCAaCgAwIBAQIQS0hw3ZuUQyKP34XGTDDdDjANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDEw1mdXNpb25hdXRoLmlvMB4XDTE5MDcyNjIyMTc0OFoXDTI5MDcyNjIyMTc0OFowGDEWMBQGA1UEAxMNZnVzaW9uYXV0aC5pbzCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAIANb2yQh0+UyDqJlA7ZJiNaysHwHlSFGFuyNoXbZWZ8JMEQ3MYc51OUGfgNZpEAV+0ufxfJXO+uDTqO5uNkGv4eslhYZH/8TkofcmQn8tRBL50S8PNjcOPTaeNe+0FeCac9UomEIKxxCKknR8cBlrZscagi9mqGxwzFIgWzg/yQYhd+aqfX9IW1f7eJWRRvDFPHBSZTZQmUXj87jCgY3dKZtVo0cTukii/LAzuPxH4hftofubctgZZDRDVd5UYN9jBDmkiL/PjiyKWMq3jX4xvkNc+kflRMET5TfkwNIka8Mje6ynA+IAWz1fXZGs/wbKWxtAUGtoZXrnPJQqhINlsCAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAL20HO6MSyDaHBQ6zELEK3RX5s0BWt/p1XcvsFvgV3YWTLbtS3GxgxsT4MC7T+2DyRQXhMWYglUFq4xyBfK4YD9rmGuG1+AF7rzmAPHZeu9JeYPR/UTkLFSHaJUp15qWYiCYM0uSEbGdD1+oXmDBSBHn/NYsx6W8XG6iIacwODr7aT231KF3JQ2Hvz+BmNuop2amxVqNBJaop/VUwfotcfPSgZzw7nkicNEzHbt/2Rm2NBsyTN0UwFyVe2GfPuTPJNEXJIPpdYElC57jHWd0PsPJComm3N93skNFVqLyMP5wXBs8Yjr/7e7c6HzRoIOFXxXZZI3ij8/XpTvWxwFwedA==
          </X509Certificate>
        </X509Data>
      </KeyInfo>
    </ns2:KeyDescriptor>
    <ns2:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://local.fusionauth.io/samlv2/login/38643239-3432-6430-3563-386530353363"/>
  </ns2:IDPSSODescriptor>
</ns2:EntityDescriptor>
```

I'll post a link to a patched jar shortly for you to test with if you'd like and then this fix will be in the upcoming release. 

Thanks!

**Comment by robotdan** [comment]
To test with the patched jar:

1. Download the ` fusionauth-samlv2-0.2.1.jar`
https://drive.google.com/file/d/1B0V3Z3UL2b_g92C1_91eyA7KBB5hOCYS/view?usp=sharing
2. Replace the existing `fusionauth-samlv2-*.jar` in the `web/WEB-INF/lib` directory. (i.e. remove the existing one, or move it somewhere else, and then copy in the new one)
3. Restart FusionAuth app service

**Comment by user1970869** [comment]
Hi,

Would you please change the permission on the linked file in the G Drive to
"anyone with the link can..." or whatever allows download without
requesting access?

Thank you!

**Comment by robotdan** [comment]
Oops, sorry. Try again. I guess "view" access isn't enough for download, I added "edit" permissions.

**Comment by robotdan** [comment]
@user1970869 were you able to download the jar and verify it worked as you expect?

**Comment by user1970869** [comment]
Patch works. Metadata can now be successfully uploaded to samltest.id

However, I can't seem to pass the IdP test [1]. I'll open a separate issue.

Thank you for the quick fix on the metadata!

**Comment by robotdan** [comment]
Available in 1.7.3
